File: engine/observability/envelope.py
"""Unified API error envelope for cross-language consistency.

所有 Go 微服务与 Python FastAPI 引擎共享同一错误响应格式：

{
    "code":      "INVALID_ARGUMENT",
    "message":   "请求参数校验失败",
    "detail":    "...",
    "trace_id":  "req-1787554500-abc123",
    "timestamp": "2026-08-27T09:30:00.123Z"
}

迁移过渡期双轨兼容：响应体同时包含 code / message / detail，
响应头强制下发 X-Request-ID 与 X-Trace-ID。

Usage::

    from engine.observability.envelope import register_envelope_exception_handlers
    register_envelope_exception_handlers(app)
"""

from __future__ import annotations

import time
from datetime import datetime, timezone
from typing import TYPE_CHECKING, Any

from fastapi import status
from fastapi.exceptions import RequestValidationError
from fastapi.responses import JSONResponse
from starlette.exceptions import HTTPException as StarletteHTTPException

# HTTP 422 constant: use int literal to avoid StarletteDeprecationWarning
# HTTP 422 常量：使用整数字面量以避免 Starlette 弃用警告
_HTTP_422 = 422

if TYPE_CHECKING:
    from fastapi import FastAPI, Request

# HTTP status code → machine-readable error code mapping.
# Aligned with Go middleware.ErrorCodeFromStatus() in pkg/middleware/envelope.go.
# HTTP 状态码 → 机器可读错误码映射，与 Go 端 ErrorCodeFromStatus() 对齐。
_STATUS_CODE_MAP: dict[int, str] = {
    400: "INVALID_ARGUMENT",
    401: "UNAUTHORIZED",
    403: "FORBIDDEN",
    404: "NOT_FOUND",
    409: "CONFLICT",
    413: "PAYLOAD_TOO_LARGE",
    422: "INVALID_ARGUMENT",
    429: "RATE_LIMITED",
    500: "INTERNAL_ERROR",
    503: "UPSTREAM_UNAVAILABLE",
}


def _get_trace_id(request: Request) -> str:
    """Extract trace ID from request headers or state.

    优先从请求头获取（上游网关/BFF 已注入），其次从 request.state 获取
    （ObservabilityMiddleware 已设置），最后自动生成。
    """
    # 1. From header (set by upstream gateway/BFF) / 从请求头获取
    trace_id = request.headers.get("X-Request-ID", "")
    if trace_id:
        return trace_id
    # 2. From request state (set by ObservabilityMiddleware) / 从请求状态获取
    if hasattr(request.state, "trace_id"):
        return str(request.state.trace_id)
    if hasattr(request.state, "request_id"):
        return str(request.state.request_id)
    # 3. Generate fallback / 自动生成
    return f"req-{int(time.time())}"


def _build_error_envelope(
    code: str,
    message: str,
    detail: Any,
    request: Request,
    status_code: int,
) -> JSONResponse:
    """Build a JSONResponse with the unified error envelope format.

    构建遵循统一错误信封格式的 JSONResponse，同时下发 X-Request-ID 与 X-Trace-ID 头。
    """
    trace_id = _get_trace_id(request)
    return JSONResponse(
        status_code=status_code,
        headers={
            "X-Request-ID": trace_id,
            "X-Trace-ID": trace_id,
        },
        content={
            "code": code,
            "message": message,
            "detail": detail,
            "trace_id": trace_id,
            "timestamp": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%f")[:-3] + "Z",
        },
    )


def _make_validation_handler(debug: bool = False):
    """Create the RequestValidationError handler with envelope format.

    创建遵循统一信封格式的请求校验异常处理器。
    debug=True 时返回详细校验错误（仅限开发环境），否则返回通用消息。
    """

    async def handler(request: Request, exc: RequestValidationError) -> JSONResponse:
        if debug:
            detail = exc.errors()
        else:
            detail = "Invalid request: one or more fields failed validation"
        return _build_error_envelope(
            code="INVALID_ARGUMENT",
            message="请求参数校验失败",
            detail=detail,
            request=request,
            status_code=_HTTP_422,
        )

    return handler


def _make_http_exception_handler():
    """Create the HTTPException handler with envelope format.

    创建遵循统一信封格式的 HTTP 异常处理器。
    将 FastAPI/Starlette HTTPException 映射为统一错误码。
    """

    async def handler(request: Request, exc: StarletteHTTPException) -> JSONResponse:
        err_code = _STATUS_CODE_MAP.get(exc.status_code, "UNKNOWN_ERROR")
        message = str(exc.detail)
        return _build_error_envelope(
            code=err_code,
            message=message,
            detail=message,
            request=request,
            status_code=exc.status_code,
        )

    return handler


def register_envelope_exception_handlers(
    app: FastAPI,
    *,
    debug_validation: bool = False,
) -> None:
    """Register unified envelope exception handlers on a FastAPI app.

    在 FastAPI 应用上注册统一信封格式的异常处理器。
    替换 main.py 中的内联异常处理器，确保所有错误响应遵循统一格式。

    Args:
        app: FastAPI application instance / FastAPI 应用实例。
        debug_validation: Return detailed validation errors (dev only)
            是否返回详细校验错误（仅开发环境使用）。
            由 PRIVACY_DEBUG_VALIDATION 环境变量控制。
    """
    app.add_exception_handler(
        RequestValidationError,
        _make_validation_handler(debug=debug_validation),
    )
    app.add_exception_handler(
        StarletteHTTPException,
        _make_http_exception_handler(),
    )

File: pkg/middleware/envelope.go
// Package middleware — unified API error envelope for cross-language consistency.
// Package middleware — 跨语言统一 API 错误信封。
//
// 所有 Go 微服务（service-hub / datasource-mgr / audit-log / bff-go / app-lz）
// 与 Python 引擎（engine/observability/envelope.py）共享同一错误响应格式：
//
//	{
//	  "code":      "INVALID_ARGUMENT",      // 机器可读错误码枚举
//	  "message":   "请求参数校验失败",        // 人读摘要
//	  "detail":    "...",                   // 兼容原 detail 字段
//	  "trace_id":  "req-1787554500-abc123", // 分布式追踪 ID
//	  "timestamp": "2026-08-27T09:30:00Z"   // UTC 时间戳
//	}
//
// 迁移过渡期双轨兼容：响应体同时包含 code / message / detail，
// 响应头强制下发 X-Request-ID 与 X-Trace-ID。
package middleware

import (
	"net/http"
	"time"

	"github.com/gin-gonic/gin"
)

// ErrorEnvelope is the unified error response body structure.
// ErrorEnvelope 统一错误响应体结构。
//
// All Go services and the Python FastAPI engine produce errors in this format,
// enabling the frontend to use a single interceptor for error parsing.
// 所有 Go 服务与 Python FastAPI 引擎均按此格式产出错误响应，
// 使前端可使用单一拦截器统一解析。
type ErrorEnvelope struct {
	Code      string `json:"code"`             // Machine-readable error code / 机器可读错误码
	Message   string `json:"message"`          // Human-readable summary / 人读摘要
	Detail    any    `json:"detail,omitempty"` // Detailed error info (optional) / 详细错误信息
	TraceID   string `json:"trace_id"`         // Distributed trace ID / 分布式追踪 ID
	Timestamp string `json:"timestamp"`        // UTC timestamp (RFC3339) / UTC 时间戳
}

// AbortWithError aborts the request and responds with a unified error envelope.
// AbortWithError 中断请求并以统一错误信封格式响应。
//
// Automatically injects X-Request-ID and X-Trace-ID response headers for
// distributed tracing correlation.
// 自动注入 X-Request-ID 与 X-Trace-ID 响应头，用于分布式追踪关联。
//
// Parameters / 参数：
//   - c: Gin context / Gin 上下文
//   - httpStatus: HTTP status code / HTTP 状态码
//   - code: Machine-readable error code (e.g. "INVALID_ARGUMENT") / 机器可读错误码
//   - message: Human-readable error summary / 人读错误摘要
//   - detail: Optional detailed error info / 可选的详细错误信息
func AbortWithError(c *gin.Context, httpStatus int, code string, message string, detail any) {
	traceID := GetTraceID(c)
	c.Header("X-Request-ID", traceID)
	c.Header("X-Trace-ID", traceID)
	c.AbortWithStatusJSON(httpStatus, ErrorEnvelope{
		Code:      code,
		Message:   message,
		Detail:    detail,
		TraceID:   traceID,
		Timestamp: time.Now().UTC().Format(time.RFC3339Nano),
	})
}

// SuccessEnvelope is the unified success response body structure (optional).
// SuccessEnvelope 统一成功响应体结构（可选使用）。
//
// For gradual migration: success responses can optionally wrap data in this
// envelope to maintain format consistency with error responses.
// 渐进迁移用：成功响应可选择性使用此信封包裹数据，保持与错误响应的格式一致性。
type SuccessEnvelope struct {
	Code      string `json:"code"`           // "OK" / 固定为 "OK"
	Message   string `json:"message"`        // Human-readable message / 人读消息
	Data      any    `json:"data,omitempty"` // Response payload / 响应数据
	TraceID   string `json:"trace_id"`       // Distributed trace ID / 分布式追踪 ID
	Timestamp string `json:"timestamp"`      // UTC timestamp / UTC 时间戳
}

// RespondWithSuccess responds with a unified success envelope.
// RespondWithSuccess 以统一成功信封格式响应。
func RespondWithSuccess(c *gin.Context, httpStatus int, message string, data any) {
	traceID := GetTraceID(c)
	c.Header("X-Request-ID", traceID)
	c.Header("X-Trace-ID", traceID)
	c.JSON(httpStatus, SuccessEnvelope{
		Code:      "OK",
		Message:   message,
		Data:      data,
		TraceID:   traceID,
		Timestamp: time.Now().UTC().Format(time.RFC3339Nano),
	})
}

// ErrorCodeFromStatus maps HTTP status codes to standard error code strings.
// ErrorCodeFromStatus 将 HTTP 状态码映射为标准错误码字串。
//
// Used by both Go services and as reference for the Python engine's code_map.
// Go 服务与 Python 引擎的 code_map 共享同一映射逻辑。
func ErrorCodeFromStatus(status int) string {
	switch status {
	case http.StatusBadRequest:
		return "INVALID_ARGUMENT"
	case http.StatusUnauthorized:
		return "UNAUTHORIZED"
	case http.StatusForbidden:
		return "FORBIDDEN"
	case http.StatusNotFound:
		return "NOT_FOUND"
	case http.StatusConflict:
		return "CONFLICT"
	case http.StatusRequestEntityTooLarge:
		return "PAYLOAD_TOO_LARGE"
	case http.StatusTooManyRequests:
		return "RATE_LIMITED"
	case http.StatusInternalServerError:
		return "INTERNAL_ERROR"
	case http.StatusServiceUnavailable:
		return "UPSTREAM_UNAVAILABLE"
	default:
		return "UNKNOWN_ERROR"
	}
}

// ExtractErrorMessage extracts the best error message from various response formats.
// ExtractErrorMessage 从多种响应格式中提取最佳错误消息。
//
// Supports: unified envelope (code+message), gin.H (detail/error), and fallback.
// 支持：统一信封（code+message）、gin.H（detail/error）及回退。
func ExtractErrorMessage(c *gin.Context, fallback string) string {
	// Try unified envelope format / 尝试统一信封格式
	if code, exists := c.Get("error_code"); exists {
		if s, ok := code.(string); ok && s != "" {
			return s
		}
	}
	// Try common error fields / 尝试常见错误字段
	for _, key := range []string{"message", "detail", "error"} {
		if val, exists := c.Get(key); exists {
			if s, ok := val.(string); ok && s != "" {
				return s
			}
		}
	}
	// Fallback to HTTP status text / 回退到 HTTP 状态文本
	if fallback != "" {
		return fallback
	}
	return http.StatusText(c.Writer.Status())
}
